### PR TITLE
Add option to serve pre-gzipped resources

### PR DESCRIPTION
**Pull request by ZizhengTai**
Changes:

- Add `preferGzipped: Boolean` parameter to `StaticFile.fromResource` to enable serving pre-gzipped files
- Add `preferGzipped: Boolean` field to `ResourceService.Config`
- Change media type for extension ".gz" from `application/x-compressed` to `application/gzip`
- Add a bunch of related tests

### CODE REVIEW

File: core/src/main/scala/org/http4s/MediaType.scala [hunk 1]
```diff
@@ -199,6 +199,7 @@ object MediaType extends Registry {
   val `application/excel`                                                         = app("excel", uncompressible, binary, "xl", "xla", "xlb", "xlc", "xld", "xlk", "xll", "xlm", "xls", "xlt", "xlv", "xlw")
   val `application/font-woff`                                                     = app("font-woff", uncompressible, binary, "woff")
   val `application/gnutar`                                                        = app("gnutar", uncompressible, binary, "tgz")
+  val `application/gzip`                                                          = app("gzip", uncompressible, binary, "gz")
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/MediaType.scala]: We should do a resynchronizing with the IANA registry at some point, but :+1: to this first step.

File: core/src/main/scala/org/http4s/StaticFile.scala [hunk 2]
```diff
@@ -126,5 +126,11 @@ object StaticFile {
     readAll[Task](f.toPath, DefaultBufferSize)
   }
 
+  private def nameToContentType(name: String): Option[`Content-Type`] =
+    name.lastIndexOf('.') match {
+      case -1 => None
+      case i => MediaType.forExtension(name.substring(i + 1)).map(`Content-Type`(_))
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/StaticFile.scala]: What happens if name is just `.`?  I can't remember whether substring out of bounds returns an empty string or throws.

Comment (ZizhengTai) [inline core/src/main/scala/org/http4s/StaticFile.scala]:
```scala
scala> ".".substring(1)
res0: String = ""
```
I think we are good.

**Comment by rossabaker** [comment]
I'll merge this by cherry-picking back to 0.16.

**Comment by rossabaker** [comment]
There are a lot of things touching StaticFile right now, and this one is a much harder cherry-pick than I anticipated.  I'm withdrawing my plan to backport this one to 0.16.x.  If another scalaz-streamer really wants it, PR welcome.

I'll merge this tomorrow barring objections.

**Pull request merged**